### PR TITLE
Use better typing in QueryBuilder

### PR DESCRIPTION
**Pull request by greg0ire**
#### Summary

That class is a bit hard to understand, so I tried to be as precise as possible while keeping both Psalm and Phpstan happy (which means I couldn't be as precise as I wanted).

### CODE REVIEW

File: lib/Doctrine/DBAL/Query/QueryBuilder.php [hunk 1]
```diff
@@ -370,7 +386,7 @@ public function setFirstResult($firstResult)
     /**
      * Gets the position of the first result the query object was set to retrieve (the "offset").
      *
-     * @return int The position of the first result.
+     * @return int|null The position of the first result.
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: Reverted to the previous situation.

File: lib/Doctrine/DBAL/Connection.php [hunk 2]
```diff
@@ -1629,8 +1629,8 @@ private function getBindingInfo($value, $type)
      * @internal This is a purely internal method. If you rely on this method, you are advised to
      *           copy/paste the code as this method may change, or be removed without prior notice.
      *
-     * @param mixed[]        $params
-     * @param int[]|string[] $types
+     * @param mixed[]                      $params
+     * @param int[]|string[]|Type[]|null[] $types
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply lib/Doctrine/DBAL/Connection.php:1633]: Yeah I asked myself if they were the same thing and tested on psalm.dev, apparently they are considered the same for now.

[the remark this answers is not in the document]
Comment (morozov) [reply lib/Doctrine/DBAL/Connection.php:1633]: They aren't the same. `a[]|b[]` is a subset of `(a|b)[]`.

[the remark this answers is not in the document]
Comment (greg0ire) [reply lib/Doctrine/DBAL/Connection.php:1633]: I know they aren't the same, and was surprised they are considered the same by both SA tools:
- https://phpstan.org/r/91010fc3-9217-4dca-9351-a2802e8ed25a
- https://psalm.dev/r/cbabf09a5b

I just elected to not also refactor this, but indeed, we might want to reconsider it at some point

Here is a preview of phpdoc we might want to look into:

<details>

```
tests/Doctrine/Tests/DBAL/Schema/IndexTest.php
107:     * @param int[]|null[] $lengths1
108:     * @param int[]|null[] $lengths2

tests/Doctrine/Tests/DBAL/Functional/Driver/PDOSqlsrv/DriverTest.php
41:     * @param int[]|string[] $driverOptions

tests/Doctrine/Tests/DBAL/Functional/Platform/DateExpressionTest.php
36:     * @return string[][]|int[][]

tests/Doctrine/Tests/DBAL/Query/Expression/ExpressionBuilderTest.php
27:     * @param string[]|CompositeExpression[] $parts
88:     * @param string[]|CompositeExpression[] $parts

tests/Doctrine/Tests/DBAL/Query/Expression/CompositeExpressionTest.php
45:     * @param string[]|CompositeExpression[] $parts

lib/Doctrine/DBAL/Connection.php
547:     * @param int[]|string[] $types  The query parameter types
564:     * @param int[]|string[] $types  The query parameter types
580:     * @param int[]|string[] $types  The query parameter types
648:     * @param int[]|string[] $types      The types of identifiers.
720:     * @param int[]|string[] $types      Types of the merged $data and $identifier arrays in that order.
755:     * @param int[]|string[] $types Types of the inserted data.
788:     * @param int[]|string[] $columnList
789:     * @param int[]|string[] $types
791:     * @return int[]|string[]
840:     * @param int[]|string[] $types  The query parameter types.
879:     * @param int[]|string[]         $types  The types the previous parameters are in.
936:     * @param int[]|string[]    $types  The types the previous parameters are in.
1565:     * @param int[]|string[]                  $types  The parameter types (PDO binding types or DBAL mapping types).
1633:     * @param int[]|string[]|Type[]|null[] $types

lib/Doctrine/DBAL/Schema/TableDiff.php
91:     * @var ForeignKeyConstraint[]|string[]

lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php
354:     * @return string[]|true[]

lib/Doctrine/DBAL/Cache/QueryCacheProfile.php
73:     * @param int[]|string[] $types

lib/Doctrine/DBAL/Query/QueryBuilder.php
109:     * @var int[]|string[]
316:     * @param int[]|string[] $types  The query parameters types to set.
353:     * @return int[]|string[] The currently defined query parameter types indexed by parameter index or name.

lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
2365:     * @param string[]|mixed[][] $definition The check definition.

lib/Doctrine/DBAL/Query/Expression/CompositeExpression.php
35:     * @var self[]|string[]
41:     * @param self[]|string[] $parts Composition of expressions to be joined on composite expression.
53:     * @param self[]|string[] $parts

lib/Doctrine/DBAL/SQLParserUtils.php
55:     * @return int[]|string[]

lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php
76:     * @var DriverConnection[]|null[]

lib/Doctrine/DBAL/Statement.php
38:     * @var int[]|string[]|Type[]|null[]

lib/Doctrine/DBAL/Sharding/ShardManager.php
69:     * @param int[]|string[] $types

```

</details>

[the remark this answers is not in the document]
Comment (morozov) [reply lib/Doctrine/DBAL/Connection.php:1633]: Agree. Let's use the old approach for now.

File: lib/Doctrine/DBAL/Query/QueryBuilder.php [hunk 3]
```diff
@@ -187,6 +202,8 @@ public function getConnection()
      * Gets the state of this query builder instance.
      *
      * @return int Either QueryBuilder::STATE_DIRTY or QueryBuilder::STATE_CLEAN.
+     *
+     * @psalm-return QueryBuilder::STATE_*
```
Comment (morozov) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php:206]: Why is this different from the type of `$state`?

File: lib/Doctrine/DBAL/Query/QueryBuilder.php [hunk 4]
```diff
@@ -1272,9 +1291,9 @@ public function __toString()
      *
      * @link http://www.zetacomponents.org
      *
-     * @param mixed  $value
-     * @param mixed  $type
-     * @param string $placeHolder The name to bind with. The string must start with a colon ':'.
+     * @param mixed                $value
+     * @param string|int|Type|null $type
+     * @param string               $placeHolder The name to bind with. The string must start with a colon ':'.
```
Comment (morozov) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php:1296]: Should it be `string|null`?

**Comment by morozov** [comment]
Given that none of the issues found during the review were found by any of the static analysis tools, what's the point of this improvement at this moment? Is it a pre-condition for say Psalm update?

**Comment by greg0ire** [comment]
> Given that none of the issues found during the review were found by any of the static analysis tools, what's the point of this improvement at this moment? Is it a pre-condition for say Psalm update?

Not at all, what prompted me to do this is #4219 , where I noticed that the code was hard to understand because of all the possibilities of the structure in `$sqlParts`. I tried to describe it accurately, but was quickly faced with hard to overcome issues, so resolved to use mixed for most key, sadly. So I'd say it has more to do with helping people than SA tools understand this class.
Do you think I should re-target to an upper branch instead?

**Comment by morozov** [comment]
Instead of beating the dead horse, why don't we backport the relevant changes from `master` (https://github.com/doctrine/dbal/pull/3830, https://github.com/doctrine/dbal/pull/3832, https://github.com/doctrine/dbal/pull/3833) to `3.0.x` and leave `2.x` alone?

**Comment by greg0ire** [comment]
Ok, I'll close this, it didn't turn out as good as I hoped anyway.

**Pull request closed without merging**